tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,373 @@
+			return err
+		}
+
+		if !keysExpired || !leaseExpired {
+			return fmt.Errorf("revoked lease %v not expired", leaseID)
+		}
+	}
+
+	ls.mu.Lock()
+	ls.revokedLeases = make(map[int64]struct{})
+	ls.mu.Unlock()
+
+	// for those leases that are alive, checks whether if they still are after waiting for 2 * TTL
+	if len(ls.aliveLeases) != 0 {
+		time.Sleep(TTL * 2 * time.Second)
+	}
+
+	plog.Printf("Alive Leases %+v ", ls.aliveLeases)
+	for leaseID := range ls.aliveLeases {
+		keysExpired, err := ls.hasKeysExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		leaseExpired, err := ls.hasLeaseExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		if keysExpired || leaseExpired {
+			return fmt.Errorf("live lease %v expired", leaseID)
+		}
+	}
+	return nil
+
+}
+
+func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.LeaseTimeToLiveResponse, error) {
+	LTL := &pb.LeaseTimeToLiveRequest{
+		ID:   leaseID,
+		Keys: true,
+	}
+	return ls.lc.LeaseTimeToLive(ctx, LTL, grpc.FailFast(false))
+}
+
+func (ls *leaseStresser) hasLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
+
+	_, err := ls.getLeaseByID(ctx, leaseID)
+
+	if err != nil {
+		switch grpc.ErrorDesc(err) {
+		case grpc.ErrorDesc(context.Canceled):
+			return false, err
+		case grpc.ErrorDesc(rpctypes.ErrGRPCLeaseNotFound):
+			plog.Printf("lease: %v is expired", leaseID)
+			return true, nil
+		}
+	}
+
+	plog.Printf("lease: %v is not expired", leaseID)
+	return false, nil
+}
+
+func (ls *leaseStresser) hasKeysExpired(ctx context.Context, leaseID int64) (bool, error) {
+	// The keys attached to the lease has the format of "<leaseID>_<idx>" where idx is the ordering key creation
+	// Since the format of keys contains about leaseID, finding keys base on "<leaseID>" prefix
+	// determines whether the attached keys for a given leaseID has been deleted or not
+	resp, err := ls.kvc.Range(ctx, &pb.RangeRequest{
+		Key:      []byte(fmt.Sprintf("%d", leaseID)),
+		RangeEnd: getPrefix([]byte(fmt.Sprintf("%d", leaseID))),
+	}, grpc.FailFast(false))
+
+	if err != nil {
+		plog.Printf("get Key info ERROR %v \n", err)
+		return false, err
+	}
+	return len(resp.Kvs) == 0, nil
+}
+
+func (ls *leaseStresser) createLeaseAndKeepAlive(ctx context.Context) (int64, error) {
+	leaseID, err := ls.createLease(ctx)
+	if err != nil {
+		plog.Printf("created lease ERROR %v", err)
+		return leaseID, err
+	}
+
+	go ls.keepLeaseAlive(ctx, leaseID)
+	return leaseID, nil
+}
+
+func (ls *leaseStresser) createLease(ctx context.Context) (int64, error) {
+	resp, err := ls.lc.LeaseGrant(ctx, &pb.LeaseGrantRequest{
+		TTL: TTL,
+	})
+
+	if err != nil {
+		plog.Printf("creating lease ERROR %v", err)
+		return 0, err
+	}
+
+	return resp.ID, nil
+}
+
+func (ls *leaseStresser) keepLeaseAlive(ctx context.Context, leaseID int64) {
+	stream, err := ls.lc.LeaseKeepAlive(ctx, grpc.FailFast(false))
+	if err != nil {
+		plog.Printf("keepLeaseAlive ERROR %v", err)
+		return
```

Comment:
I want to kill the keep alive go routine if I can't create a stream. not sure how i can get around of that elegantly without return or break. I could do the following

``` go
stream, err := ls.lc.LeaseKeepAlive(ctx, grpc.FailFast(false))
	if err == nil {
            for {
		select {
		case <-time.After(1 * time.Second):
		case <-stream.Context().Done():
			plog.Printf("Stresser %v keepLeaseAlive stream closed", ls.id)
			return
		}
		err = stream.Send(&pb.LeaseKeepAliveRequest{ID: leaseID})
		if err != nil {
			plog.Printf("keepLeaseAlive stream send lease %v ERROR %v ", leaseID, err)
			continue
		}

		respRC, err := stream.Recv()

		if err != nil {
			plog.Printf("keepLeaseAlive stream recieve lease %v stream ERROR %v ", leaseID, err)
			continue
		}
		// lease expires after TTL become 0
		// don't send keepalive if the lease has expired
		if respRC.TTL <= 0 {
			plog.Printf("keepLeaseAlive: lease %v has expired ", respRC.ID)
			return
		}
	}

       plog.Printf("keepLeaseAlive ERROR %v", err)
```